prow/kube/prowjob_collector.go

```diff
@@ -0,0 +1,253 @@
+		lister:   lister,
+		selector: selector,
+		agent:    agent,
+	}
+	prometheus.WrapRegistererWith(prometheus.Labels{"collector_name": name}, reg).MustRegister(c)
+	return c
+}
+
+var (
+	invalidLabelCharRE    = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+	escapeWithDoubleQuote = strings.NewReplacer("\\", `\\`, "\n", `\n`, "\"", `\"`)
+)
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/internal/store/utils.go#L60
+func kubeLabelsToPrometheusLabels(labels map[string]string) ([]string, []string) {
+	labelKeys := make([]string, 0, len(labels))
+	for k := range labels {
+		labelKeys = append(labelKeys, k)
+	}
+	sort.Strings(labelKeys)
+
+	labelValues := make([]string, 0, len(labels))
+	for i, k := range labelKeys {
+		labelKeys[i] = "label_" + sanitizeLabelName(k)
+		labelValues = append(labelValues, escapeString(labels[k]))
+	}
+	return labelKeys, labelValues
+}
+
+func sanitizeLabelName(s string) string {
+	return invalidLabelCharRE.ReplaceAllString(s, "_")
+}
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/pkg/metric/metric.go#L96
+func escapeString(v string) string {
+	return escapeWithDoubleQuote.Replace(v)
+}
+
+const (
+	pjControllerName = "prowjob-controller-for-collector"
+)
+
+type PJController struct {
+	lister prowjoblisterv1.ProwJobLister
+	queue  workqueue.RateLimitingInterface
+	logger *logrus.Entry
+}
+
+func NewPJController(informer prowjobinformerv1.ProwJobInformer, client prowjobset.Interface) *PJController {
+	logger := logrus.WithField("controller", pjControllerName)
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logger.Infof)
+	eventBroadcaster.StartRecordingToSink(&coreclient.EventSinkImpl{Interface: coreclient.New(client.ProwV1().RESTClient()).Events("")})
+
+	c := &PJController{
+		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), pjControllerName),
+		logger: logger,
+		lister: informer.Lister(),
+	}
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.add,
+		UpdateFunc: c.update,
+	})
+
+	return c
+}
+
+func (c *PJController) add(obj interface{}) {
+	prowJob := obj.(*prowapi.ProwJob)
+	c.logger.Debugf("enqueueing added prow job %s/%s", prowJob.GetNamespace(), prowJob.GetName())
+	c.enqueue(prowJob)
+}
+
+func (c *PJController) update(old, obj interface{}) {
+	prowJob := obj.(*prowapi.ProwJob)
+	c.logger.Debugf("enqueueing updated prow job %s/%s", prowJob.GetNamespace(), prowJob.GetName())
+	c.enqueue(prowJob)
+}
+
+func (c *PJController) enqueue(obj metav1.Object) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *PJController) List(opts metav1.ListOptions) ([]*prowapi.ProwJob, error) {
+	var pjs []*prowapi.ProwJob
+	for c.processNextWorkItem(&pjs) {
```

Comment:
@stevekuznetsov 

I just realized ... informer generates the events, and puts them into a queue and lister takes them out. A lister acts as a worker of the queue. Http calls from prometheus server scraping activate our workers.

So I got the feeling that ... say we have 2 prometheus servers and it means they could get different scaping samples from plank ... this is not nice at all (our BLOCKER here?).

Maybe the queue is not a proper data structure here and that is my missing part of the logic in `kube-state-metrics`.